page-tracking/src/page_tracker.rs

```diff
@@ -298,6 +324,55 @@ impl PageTracker {
         }
     }
 
+    /// Acquires an exclusive reference to the shareable page at `addr` if it's owned by owner,
+    /// and in Mapped or Shared state
+    pub fn get_shareable_page<P: PhysPage>(
+        &self,
+        addr: SupervisorPageAddr,
+        owner: PageOwnerId,
+    ) -> Result<P> {
+        let mut page_tracker = self.inner.lock();
+        let info = page_tracker.get_mut(addr)?;
+        if info.owner() != Some(owner) || info.mem_type() != P::mem_type() {
+            return Err(Error::PageNotShareable);
+        }
+
+        match info.state() {
+            // Ensure that the share RC can be incremented before attempting to obtain the lock.
+            PageState::Shared(rc) if rc == u64::MAX => {
+                return Err(Error::PageNotShareable);
+            }
+            PageState::Shared(_) | PageState::Mapped => {}
+            _ => {
+                return Err(Error::PageNotShareable);
+            }
+        };
+
+        info.lock_page()?;
+        // Safe since we've taken exclusive ownership of the page, verified its typing.
+        // TODO: Page size
+        Ok(unsafe { P::new(addr) })
+    }
+
+    /// Bumps the reference count on a the shareable page at `addr` if it's owned by owner, in Mapped or Shared state, and has a
+    /// current reference count < u64::MAX. The exclusive reference to the page is released following a successful share.
+    pub fn share_page(
+        &self,
+        addr: SupervisorPageAddr,
```

Comment:
I've been trying to avoid `PageTracker` APIs that operate on raw addresses and instead have been using `PhysPage` when possible since that implies exclusive ownership, whereas a raw address type does not imply anything about ownership. Can we make this take, say, a `ShareablePhysPage` and return a `MappablePhysPage` in a similar manner to `reclaim_page()` or `assign_page_*()`?

Reply:
Changed.